text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,676 @@
+blocker for making progress on this issue. Stabilizing the macro will unblock
+projects that wish to define custom smart pointers, and does not prevent
+evolution of the underlying traits.
+
+See also [the section on prior art][prior-art], which discusses a previous
+attempt to stabilize the underlying traits.
+
+## Receiver and Deref traits
+
+The vast majority of custom smart pointers will implement `Receiver` (often via
+`Deref`, which results in a `Receiver` impl due to the blanket impl). So why
+not also emit a `Receiver`/`Deref` impl in the output of the macro. One
+advantage of doing so is that this may sufficiently limit the macro so that we
+do not need to solve the pin soundness issue discussed in [the unresolved
+questions section][unresolved-questions].
+
+However, it turns out that there are quite a few different ways we might
+implement `Deref`. For example, consider [the custom `Rc` example][custom-rc]:
+```rust
+#[derive(SmartPointer)]
+pub struct Rc<T: ?Sized> {
+    inner: NonNull<RcInner<T>>,
+}
+
+struct RcInner<T: ?Sized> {
+    refcount: usize,
+    value: T,
+}
+
+impl<T: ?Sized> Deref for Rc<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        let ptr = self.inner.as_ptr();
+        unsafe { &*ptr.value }
+    }
+}
+```
+Making the macro general enough to generate `Deref` impls that are _that_
+complex would not be feasible. And it doesn't make sense to stabilize the macro
+without support for the custom `Rc` case, as implementing a custom `Arc` in the
+Linux Kernel is the primary motivation for this RFC.
+
+Note that having the macro generate a `Receiver` impl instead doesn't work
+either, because that prevents the user from implementing `Deref` at all. (There
+is a blanket impl of `Receiver` for all `Deref` types.)
+
+## Why not two derive macros?
+
+The derive macro generates two different trait implementations:
+
+- [`CoerceUnsized`] that allows conversions from `SmartPtr<MyStruct>` to
+  `SmartPtr<dyn MyTrait>`.
+- [`DispatchFromDyn`] that allows conversions from `SmartPtr<dyn MyTrait>` to
+  `SmartPtr<MyStruct>`.
+
+It could be argued that these should be split into two separate derive macros.
+We are not proposing this for a few reasons:
+
+- If there are two derive macros, then we have to support the case where you
+  only use one of them. There are use-cases for implementing [`CoerceUnsized`]
+  without [`DispatchFromDyn`], but you do this for cases where your type is not
+  a smart pointer, but rather a transparent container like [`Cell`]. It makes
+  coercions like `Cell<Box<MyStruct>>` to `Cell<Box<dyn MyTrait>>` possible.
+  Supporting that is a significantly increased scope of the RFC, and the
+  authors believe that supporting transparent containers should be a separate
+  follow-up RFC.
+
+- Right now there are use cases for `CoerceUnsized` (transparent containers)
+  and `CoerceUnsized+DispatchFromDyn` (smart pointers), but there aren't any
```

Comment:
actually `Cell` also [implemented](https://doc.rust-lang.org/std/ops/trait.DispatchFromDyn.html#impl-DispatchFromDyn%3CCell%3CU%3E%3E-for-Cell%3CT%3E) `DispatchFromDyn`.

Reply:
I see. Then I don't think I know of any use-cases for implementing one but not both of the traits. Unless you count causes where the `where` clauses differ such as https://github.com/rust-lang/rfcs/pull/3621#discussion_r1587353825.